### PR TITLE
Make AuthToken model pluggable [WIP]

### PR DESCRIPTION
**Pull request by kalekseev**
Partially fixes https://github.com/James1345/django-rest-knox/issues/159.

This is a proof of concept that the feature is not that complex and can be added with minimal breaking changes. Would like to receive comments on that before investing more time.

### CODE REVIEW

**Comment by johnraz** [comment]
Hi! Thanks a lot for this. The history seems to be a bit messed up. Could you rebase and make sure only your commits appear. I’ll try to follow up with a review ASAP once that’s done.

**Comment by kalekseev** [comment]
@johnraz done

**Comment by kalekseev** [comment]
it seems that develop branch doesn't contain all the master commits so pr includes a couple of comments from master

**Comment by johnraz** [comment]
You should base your work on develop, not master.

File: knox/auth.py [hunk 1]
```diff
@@ -34,7 +34,9 @@ class TokenAuthentication(BaseAuthentication):
     - `request.user` will be a django `User` instance
     - `request.auth` will be an `AuthToken` instance
     '''
-    model = AuthToken
+
+    def get_model(self):
+        return get_token_model()
```
Comment (johnraz) [inline knox/auth.py]: I would prefer to have this declared on the top of the file, in the same fashion the `User` model is declared.
What would be the use case to override this method?

File: knox/auth.py [hunk 2]
```diff
@@ -62,7 +64,8 @@ def authenticate_credentials(self, token):
         '''
         msg = _('Invalid token.')
         token = token.decode("utf-8")
-        for auth_token in AuthToken.objects.filter(
+        model = self.get_model()
+        for auth_token in model.objects.filter(
```
Comment (johnraz) [inline knox/auth.py]: This implies that the new "AuthToken" model should have a field `token_key`, I'm wondering if we should not make this more flexible and have a `get_tokens` method has another way to fetch back the token.

File: knox/models.py [hunk 3]
```diff
@@ -19,7 +19,7 @@ def create(self, user, expiry=knox_settings.TOKEN_TTL):
 
         instance = super(AuthTokenManager, self).create(
             token_key=token[:CONSTANTS.TOKEN_KEY_LENGTH], digest=digest,
-            salt=salt, user=user, expiry=expiry)
+            salt=salt, user=user, expiry=expiry, **kwargs)
```
Comment (johnraz) [inline knox/models.py]: Can you explain this change?

File: knox/models.py [hunk 4]
```diff
@@ -38,5 +38,10 @@ class AuthToken(models.Model):
     created = models.DateTimeField(auto_now_add=True)
     expiry = models.DateTimeField(null=True, blank=True)
 
+    class Meta:
+        # Work around for a bug in Django:
+        # https://code.djangoproject.com/ticket/19422
+        abstract = 'knox' not in settings.INSTALLED_APPS
```
Comment (johnraz) [inline knox/models.py]: Reading the ticket, this seems to only apply to pre "app loading" times in Django. I don't think we care.

File: knox/views.py [hunk 5]
```diff
@@ -60,7 +60,7 @@ def post(self, request, format=None):
                     status=status.HTTP_403_FORBIDDEN
                 )
         token_ttl = self.get_token_ttl()
-        instance, token = AuthToken.objects.create(request.user, token_ttl)
+        instance, token = get_token_model().objects.create(request.user, token_ttl)
```
Comment (johnraz) [inline knox/views.py]: As per my comment above, this still implies some constraint on how the custom AuthToken model would work as it needs to comply to what happens in the create method and also reuse the manager from the default model.

File: knox/models.py (hunk 4 above)
Comment (kalekseev) [reply knox/models.py]: Basically the problem is following, if end user want to inherit from knox.models.AuthToken then he will get `RuntimeError: Model class knox.models.AuthToken doesn't declare an explicit app_label and isn't in an application in INSTALLED_APPS.`

File: knox/auth.py (hunk 2 above)
Comment (johnraz) [reply knox/auth.py]: Yes sure but still I was challenging the fact that we could make this even more flexible in the case someone needs it. If the point is to make the model pluggable, I would probably go for something fully pluggable unless it's proven too complex.

File: knox/models.py (hunk 4 above)
Comment (johnraz) [reply knox/models.py]: I still have the feeling that this is not relevant anymore, did you witness the problem on a recent / supported version of Django?

File: knox/views.py (hunk 5 above)
Comment (johnraz) [reply knox/views.py]: That or we could simply have a `create_token(request)` method.

File: knox/models.py (hunk 3 above)
Comment (johnraz) [reply knox/models.py]: Ok but then you need to allow those params to be passed everywhere create is used in the code and this is not the case right now.

**Comment by kalekseev** [comment]
@alex-yobota sorry but my original plan is different from what maintainers want:

> Yes sure but still I was challenging the fact that we could make this even more flexible in the case someone needs it. If the point is to make the model pluggable, I would probably go for something fully pluggable unless it's proven too complex.

It's quite reasonable decision to make it fully customizable but I don't have resources to do that now. If I will have new projects with similar requirements in near future I will implement this.

**Comment by alex-yobota** [comment]
> @alex-yobota sorry but my original plan is different from what maintainers want:
> 
> > Yes sure but still I was challenging the fact that we could make this even more flexible in the case someone needs it. If the point is to make the model pluggable, I would probably go for something fully pluggable unless it's proven too complex.
> 
> It's quite reasonable decision to make it fully customizable but I don't have resources to do that now. If I will have new projects with similar requirements in near future I will implement this.

Cool, thanks for trying anyway :)

File: knox/models.py (hunk 4 above)
Comment (yotamgod) [reply knox/models.py:44]: I'm not sure this is what you mean, but I can confirm that on Django 2.2 this (or the lack of this) causes issues making knox less plug-able.
For example, when inheriting from AuthToken and adding field to the model, creating a multi-column constraint (unique_together) is not possible (there are two tables).

Comment (kalekseev) [reply knox/models.py:44]: yes, that the same issue that DRF auth token had, so I just copied their solution here

**Comment by johnraz** [comment]
As this PR is quite old and it was decided to make this more pluggable I'm going to close for now.

**Pull request closed without merging**